File: frontend/src/api/jobs.ts
import apiClient from './client'

export type JobStatus = 'draft' | 'open' | 'paused' | 'closed'

export interface StructuredJd {
  responsibilities: string[]
  requirements: string[]
  must_have: string[]
  nice_to_have: string[]
  skills: string[]
}

export interface ScoreDimension {
  name: string
  weight: number
  description: string
  anchors: string[]
  custom?: boolean
}

export interface JobVersion {
  id: string
  version_label: string | null
  major: number
  minor: number
  status: string
  upgrade_type: string | null
  change_summary: string | null
  raw_jd_text: string
  structured_jd: StructuredJd
  score_dimensions: ScoreDimension[]
  job_snapshot?: Record<string, unknown> | null
  base_version_id: string | null
  published_at: string | null
  published_by: string | null
  created_at: string
  updated_at: string
  is_current?: boolean
  bound_candidates?: number
}

export interface JobVersionListItem {
  id: string
  version_label: string | null
  major: number
  minor: number
  status: string
  status_label: string
  upgrade_type: string | null
  upgrade_type_label: string | null
  change_summary: string | null
  published_at: string | null
  published_by: string | null
  created_at: string
  updated_at: string
  is_current: boolean
  is_draft: boolean
  bound_candidates: number
}

export interface VersionDiffChange {
  field: string
  label: string
  before: unknown
  after: unknown
}

export interface VersionDiffResponse {
  from_version: JobVersionListItem
  to_version: JobVersionListItem
  changes: VersionDiffChange[]
  has_changes: boolean
}

export interface JobListItem {
  id: string
  code: string
  status: JobStatus
  status_label: string
  name: string
  department: string
  level: string | null
  headcount: number | null
  location: string
  owner_user_id: string | null
  owner_name: string
  urgency: string | null
  current_version_label: string | null
  updated_at: string
  created_at: string
}

export interface JobListResponse {
  items: JobListItem[]
  total: number
  page: number
  page_size: number
}

export interface JobDetail {
  id: string
  code: string
  status: JobStatus
  status_label: string
  name: string
  department: string
  level: string | null
  headcount: number | null
  location: string
  owner_user_id: string | null
  owner_name: string
  urgency: string | null
  source_job_id: string | null
  current_version_id: string | null
  draft_version_id: string | null
  current_version: JobVersion | null
  draft_version: JobVersion | null
  closed_at: string | null
  close_reason: string | null
  pause_reason: string | null
  created_by: string | null
  updated_by: string | null
  created_at: string
  updated_at: string
}

export interface CreateJobPayload {
  name?: string
  department?: string
  level?: string | null
  headcount?: number | null
  location?: string
  owner_name?: string
  urgency?: string | null
  raw_jd_text?: string
  structured_jd?: StructuredJd
  score_dimensions?: ScoreDimension[]
}

export interface SaveDraftPayload extends CreateJobPayload {
  change_summary?: string | null
}

export interface JobListQuery {
  page?: number
  page_size?: number
  code?: string
  name?: string
  keyword?: string
  department?: string
  owner?: string
  status?: JobStatus | ''
}

export function emptyStructuredJd(): StructuredJd {
  return {
    responsibilities: [],
    requirements: [],
    must_have: [],
    nice_to_have: [],
    skills: [],
  }
}

export async function listJobs(params: JobListQuery = {}) {
  const { data } = await apiClient.get<JobListResponse>('/jobs', { params })
  return data
}

export async function getJob(jobId: string) {
  const { data } = await apiClient.get<JobDetail>(`/jobs/${jobId}`)
  return data
}

export async function createJob(payload: CreateJobPayload) {
  const { data } = await apiClient.post<JobDetail>('/jobs', payload)
  return data
}

export async function saveJobDraft(jobId: string, payload: SaveDraftPayload) {
  const { data } = await apiClient.patch<JobDetail>(`/jobs/${jobId}/draft`, payload)
  return data
}

export async function publishJob(
  jobId: string,
  payload: { reason?: string; force_major?: boolean; change_summary?: string } = {},
) {
  const { data } = await apiClient.post<JobDetail>(`/jobs/${jobId}/publish`, payload)
  return data
}

export async function pauseJob(jobId: string, reason: string) {
  const { data } = await apiClient.post<JobDetail>(`/jobs/${jobId}/pause`, { reason })
  return data
}

export async function resumeJob(jobId: string, reason: string) {
  const { data } = await apiClient.post<JobDetail>(`/jobs/${jobId}/resume`, { reason })
  return data
}

export async function closeJob(jobId: string, reason: string) {
  const { data } = await apiClient.post<JobDetail>(`/jobs/${jobId}/close`, { reason })
  return data
}

export async function copyJob(jobId: string) {
  const { data } = await apiClient.post<JobDetail>(`/jobs/${jobId}/copy`)
  return data
}

export async function deleteJob(jobId: string) {
  await apiClient.delete(`/jobs/${jobId}`)
}

export async function listJobVersions(jobId: string) {
  const { data } = await apiClient.get<{ items: JobVersionListItem[]; total: number }>(
    `/jobs/${jobId}/versions`,
  )
  return data
}

export async function getJobVersion(jobId: string, versionId: string) {
  const { data } = await apiClient.get<JobVersion>(`/jobs/${jobId}/versions/${versionId}`)
  return data
}

export async function diffJobVersions(
  jobId: string,
  fromVersionId: string,
  toVersionId: string,
) {
  const { data } = await apiClient.get<VersionDiffResponse>(`/jobs/${jobId}/versions/diff`, {
    params: { from: fromVersionId, to: toVersionId },
  })
  return data
}

export async function copyVersionToDraft(jobId: string, versionId: string) {
  const { data } = await apiClient.post<JobDetail>(
    `/jobs/${jobId}/versions/${versionId}/copy-to-draft`,
  )
  return data
}

File: backend/app/repositories/jobs.py
from copy import deepcopy
from datetime import UTC, datetime
from uuid import UUID

from sqlalchemy import func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.job import (
    JOB_STATUS_DRAFT,
    VERSION_STATUS_DRAFT,
    Job,
    JobCodeSequence,
    JobVersion,
    empty_structured_jd,
)


class JobNotFoundError(Exception):
    pass


async def allocate_job_code(
    session: AsyncSession,
    *,
    now: datetime | None = None,
) -> str:
    current = now or datetime.now(UTC)
    year_month = current.strftime("%Y%m")
    seq = await session.scalar(
        select(JobCodeSequence)
        .where(JobCodeSequence.year_month == year_month)
        .with_for_update()
    )
    if seq is None:
        seq = JobCodeSequence(year_month=year_month, last_value=0)
        session.add(seq)
        await session.flush()
    seq.last_value += 1
    await session.flush()
    return f"JOB-{year_month}-{seq.last_value:04d}"


async def get_job_by_id(session: AsyncSession, job_id: UUID) -> Job | None:
    return await session.scalar(
        select(Job)
        .options(selectinload(Job.versions))
        .where(Job.id == job_id)
    )


async def list_jobs(
    session: AsyncSession,
    *,
    page: int,
    page_size: int,
    code: str | None = None,
    name: str | None = None,
    keyword: str | None = None,
    department: str | None = None,
    owner: str | None = None,
    status: str | None = None,
    updated_from: datetime | None = None,
    updated_to: datetime | None = None,
) -> tuple[list[Job], int]:
    query = select(Job).options(selectinload(Job.versions))
    count_query = select(func.count()).select_from(Job)

    if keyword:
        pattern = f"%{keyword}%"
        clause = or_(
            Job.code.ilike(pattern),
            Job.name.ilike(pattern),
            Job.department.ilike(pattern),
        )
        query = query.where(clause)
        count_query = count_query.where(clause)
    if code:
        query = query.where(Job.code.ilike(f"%{code}%"))
        count_query = count_query.where(Job.code.ilike(f"%{code}%"))
    if name:
        query = query.where(Job.name.ilike(f"%{name}%"))
        count_query = count_query.where(Job.name.ilike(f"%{name}%"))
    if department:
        query = query.where(Job.department.ilike(f"%{department}%"))
        count_query = count_query.where(Job.department.ilike(f"%{department}%"))
    if owner:
        query = query.where(Job.owner_name.ilike(f"%{owner}%"))
        count_query = count_query.where(Job.owner_name.ilike(f"%{owner}%"))
    if status:
        query = query.where(Job.status == status)
        count_query = count_query.where(Job.status == status)
    if updated_from is not None:
        query = query.where(Job.updated_at >= updated_from)
        count_query = count_query.where(Job.updated_at >= updated_from)
    if updated_to is not None:
        query = query.where(Job.updated_at <= updated_to)
        count_query = count_query.where(Job.updated_at <= updated_to)

    total = await session.scalar(count_query) or 0
    result = await session.scalars(
        query.order_by(Job.updated_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
    )
    return list(result.all()), int(total)


def get_version_by_id(job: Job, version_id: UUID | None) -> JobVersion | None:
    if version_id is None:
        return None
    for version in job.versions:
        if version.id == version_id:
            return version
    return None


async def create_job_with_draft(
    session: AsyncSession,
    *,
    code: str,
    actor_id: UUID,
    name: str = "",
    department: str = "",
    level: str | None = None,
    headcount: int | None = None,
    location: str = "",
    owner_user_id: UUID | None = None,
    owner_name: str = "",
    urgency: str | None = None,
    source_job_id: UUID | None = None,
    raw_jd_text: str = "",
    structured_jd: dict | None = None,
    score_dimensions: list | None = None,
) -> Job:
    job = Job(
        code=code,
        status=JOB_STATUS_DRAFT,
        name=name,
        department=department,
        level=level,
        headcount=headcount,
        location=location,
        owner_user_id=owner_user_id,
        owner_name=owner_name,
        urgency=urgency,
        source_job_id=source_job_id,
        created_by=actor_id,
        updated_by=actor_id,
    )
    session.add(job)
    await session.flush()

    version = JobVersion(
        job_id=job.id,
        version_label="draft",
        major=0,
        minor=0,
        status=VERSION_STATUS_DRAFT,
        upgrade_type=None,
        raw_jd_text=raw_jd_text,
        structured_jd=deepcopy(structured_jd or empty_structured_jd()),
        score_dimensions=deepcopy(score_dimensions or []),
        created_by=actor_id,
    )
    session.add(version)
    await session.flush()
    job.draft_version_id = version.id
    await session.flush()
    await session.refresh(job, attribute_names=["versions"])
    return job


async def create_draft_version_from_base(
    session: AsyncSession,
    *,
    job: Job,
    base: JobVersion,
    actor_id: UUID,
) -> JobVersion:
    version = JobVersion(
        job_id=job.id,
        version_label="draft",
        major=0,
        minor=0,
        status=VERSION_STATUS_DRAFT,
        upgrade_type=None,
        raw_jd_text=base.raw_jd_text,
        structured_jd=deepcopy(base.structured_jd or empty_structured_jd()),
        score_dimensions=deepcopy(base.score_dimensions or []),
        job_snapshot=deepcopy(base.job_snapshot) if base.job_snapshot else None,
        base_version_id=base.id,
        created_by=actor_id,
    )
    session.add(version)
    await session.flush()
    job.draft_version_id = version.id
    await session.flush()
    return version


async def delete_job(session: AsyncSession, job: Job) -> None:
    await session.delete(job)
    await session.flush()
